Topics/Tech_Stacks/Unity_ML_Agents.md

```diff
@@ -0,0 +1,199 @@
+
+```public override void OnActionReceived(ActionBuffers actions)```
+
+```public void OnTriggerEnter(Collider other)```
+
+In the following few sections, we will go through and explain each of these functions in detail. We will first begin with ```OnActionReceieved()```
+
+Unlike us, our agent is only able to play through making actions through numbers. These numbers will be given through the ActionBuffer. As specified before, we want our agent to be able to move left, right, and stand still. We will accomplish this by assigning each action a number. We begin by obtaining the action from DiscreteActions, or integer actions. We will set up where our AI gets these DiscreteActions later in our observations section.
+
+```
+public override void OnActionReceived(ActionBuffers actions)
+    {
+        int action = actions.DiscreteActions[0];
+        switch (action)
+        {
+            case 0:
+                transform.Translate(Vector3.left * Time.deltaTime);
+                break;
+            case 1:
+                transform.Translate(Vector3.right * Time.deltaTime);
+                break;
+            case 2:
+                break;
+        }
+    }
+```
+While it might make sense for us to assign case 0 to moving left, and either case 1 or 2 to moving right, the agent does not have any context to what each number means. This means that any number can be assigned to any action, and they will all be treated equally by our agent.
+
+Now that our agent is capable of making moves, we will now make it able to make observations.
+
+# Observations
+
+In order for our agent to learn from its environment, it must be able to gather information from its environment. As our agent only needs to solve a very simple problem, we will give it some very simple information. To solve this problem, it should be able to observe its own position, and the position of both rewards.
+
+To make things easier, we should add a reference to both rewards at the start of our agent class
+
+```
+public class CollectRewardAgent : Agent
+{
+    public GameObject rewardSmall;
+    public GameObject rewardLarge;
+    ...
+```
+
+This will allow us to easily access the position of both reward objects as well. To add an observation, we simply go under the CollectObservations function and call ```Sensor.AddObservation()```. We would like to add the x value of all 3 positions, as our agent will only be capable of moving along the x axis. 
+
+```
+public override void CollectObservations(VectorSensor sensor)
+    {
+        sensor.AddObservation(rewardSmall.transform.localPosition.x);
+        sensor.AddObservation(rewardLarge.transform.localPosition.x);
+        sensor.AddObservation(transform.localPosition.x);
+    }
+```
+
+Here we are adding the local positions of all 3 objects. I use local position here so that if we move our training setup around, everything stays functional. However, you can adjust this depending on how your training setup may look like.
```

Comment:
First time you used "I" instead of "we", you should change it to keep it consistent

Reply:
I originally used I to differentiate how this is something that I did for the project, but was mostly optional. 
To avoid confusion, I replaced it with we and reworded my explanation for why I used local positioning